interfaces/cython/cantera/func1.pyx

```diff
@@ -3,6 +3,13 @@
 
 import sys
 
+
+# list holding references to Python objects used for Func1 calls;
+# prevents garbage collection of temporary Python objects, which would otherwise
+# lead to segfaults when C++ Func1 objects are evaluated
+cdef list _python_object_registry = []
```

Comment:
Is this going to leak memory when `Func1` objects go out of scope and _should_ be garbage collected? I don't understand quite enough about the problem to suggest an alternate approach, but just a thought I had looking through this

Reply:
Yep, I got the purpose. I'm wondering what happens if someone creates lots of objects which aren't going to be used any more. Those should be garbage collected but this will retain a reference to them indefinitely, and not release that memory back. Of course when the Python process ends, the memory will be released so it's not a true memory leak.

However, I'm not sure how to tell when a Func1 object won't be used anymore... Or rather, I'm not sure why Python is deciding to GC the C++ object (that's what's happening, right?). So I'm not sure of a better solution either at the moment.

Reply:
No - it garbage collects the Python object that C++ still wants to call - any ideas would be appreciated. As an aside, I dropped some of the newer commits as this is probably enough to review by itself.

Reply:
One more thing: `CustomRate` actually holds a reference to the Python `Func1` in Cython, but apparently Python isn't aware of that and still thinks that it's unused, see https://github.com/Cantera/cantera/blob/0925fc08634515458bd2feb018ca29888f8aecdd/interfaces/cython/cantera/_cantera.pxd#L1360-L1362

Reply:
Hah! Nevermind - found a way around that. GC does work as intended, but not if the Python `Reaction` objects are not kept around.